### PR TITLE
CLOUDP-238528: Atlas stream connection controller

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: pkg/controller/atlasstream/atlasstream_connection_controller.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	err := r.Client.List(ctx, streamInstances, listOps)
+	if err != nil {
+		return r.terminate(workflowCtx, workflow.Internal, err)
+	}
+
+	if len(streamInstances.Items) > 0 {
+		return r.lock(workflowCtx, akoStreamConnection)
+	}
+
+	return r.release(workflowCtx, akoStreamConnection)
```
Comment (s-urbaniak) [inline pkg/controller/atlasstream/atlasstream_connection_controller.go:91]: i really like the `release` and `lock` names for those transitions, great job! 👏

**Comment by helderjs** [comment]
> @helderjs do you mind to draw a state machine diagram here as well? it should be mechanical exercise based on the state machine here 👍

Sure! I'll put on the spec as done with the instances

File: pkg/controller/atlasstream/atlasstream_connection_controller.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+			},
+		)
+	}
+
+	return requests
+}
+
+func (r *AtlasStreamsConnectionReconciler) skip(ctx context.Context, log *zap.SugaredLogger, streamConnection *akov2.AtlasStreamConnection) ctrl.Result {
+	log.Infow(fmt.Sprintf("-> Skipping AtlasStreamConnection reconciliation as annotation %s=%s", customresource.ReconciliationPolicyAnnotation, customresource.ReconciliationPolicySkip), "spec", streamConnection.Spec)
+	if !streamConnection.GetDeletionTimestamp().IsZero() {
```
Comment (s-urbaniak) [inline pkg/controller/atlasstream/atlasstream_connection_controller.go:129]: this case is missing unit test coverage

File: pkg/controller/atlasstream/atlasstream_connection_controller.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+	akoStreamConnection := akov2.AtlasStreamConnection{}
+	result := customresource.PrepareResource(ctx, r.Client, req, &akoStreamConnection, log)
+	if !result.IsOk() {
+		return result.ReconcileResult(), nil
+	}
+
+	return r.ensureAtlasStreamConnection(ctx, log, &akoStreamConnection)
+}
+
```
Comment (josvazg) [inline pkg/controller/atlasstream/atlasstream_connection_controller.go:57]: Maybe a comment of the state machine here?

> - check preconditions ...
> ...
> - if found lock it with a finalizer
> - if not release by clearing the finalizer

**Review by roothorp — APPROVED** [review]

**Review by igor-karpukhin — APPROVED** [review]

**Pull request merged**